Feature/create group endpoint

Levliv opened the pull request:
http://outsourcing.nat.tepkom.ru:20080/redmine/issues/1135
* Create GroupController
* Create Create endpoint for GroupController
* Create Command with business login for Create Group endpoint
* CreateGroupRequest
* CreateGroupQuestionRequest
* CreateGroupRequestValidator
* CreateGroupQuestionRequestValidator
* DbGroupRequestMapper

Levliv committed:
+CreateGroupCommand(Execute async), +CreateGroupRequest, +Validator to be finished, +DbGroupMapper(Map), +GroupRepository (CreateGroupAsync)

Levliv committed:
Amended CreateGroupRequest.cs: Refactor QuestionCollection,
Amended CreateGroupRequestValidator.cs: +Checking group.Subject, group.Description, group.Questions,
Created CreateQuestionRequest.cs,
Amended DbGroupMapper.cs: +Mapping group.Questions.

Levliv committed:
Minor style changes in CreateGroupCommand.cs,
+GroupController

Levliv committed:
Amended CreateGroupRequestValidator.cs: +CreateQuestionRequestValidator

Levliv committed:
Delete launchSettings.json

Levliv committed:
Some style changes

Levliv committed:
Merge remote-tracking branch 'origin/feature/CreateGroup_Endpoint' into feature/CreateGroup_Endpoint

# Conflicts:
#	src/SurveyService/Properties/launchSettings.json

Levliv committed:
Minor style changes

Levliv committed:
Deleted odd files

Levliv committed:
Edited validator

Levliv committed:
Deleted useless using

VLiamin commented on src/SurveyService.Business/Commands/Group/CreateGroupCommand.cs:
```diff
@@ -0,0 +1,60 @@
+      return _responseCreator.CreateFailureResponse<Guid>(
+        HttpStatusCode.BadRequest,
+        new List<string> { "Error occured while creating group." }
+        );
```
May be replace it to line before?

Levliv replied:
done

VLiamin commented on src/SurveyService/Controllers/GroupController.cs:
```diff
@@ -0,0 +1,21 @@
+{
+  [HttpPost("create")]
+  public async Task<OperationResultResponse<Guid>> CreateAsync(
+    ICreateGroupCommand command,
```
May be add attribute [FromServices]?

Levliv replied:
Trying to be consistent with other code, DI will work without that.

Levliv replied:
done

Levliv replied:
In other controllers you use the attribute, but DI would have worked without that?
@vorobevamv

vorobevamv replied:
I'm not sure. Here can be a lot of sources (services, body, query). Even if it works, it is difficult to read code.
You can try to run without attributes, then tell us))

VLiamin commented on src/SurveyService.Validation/Question/Interfaces/ICreateQuestionRequestValidator.cs:
```diff
@@ -0,0 +1,11 @@
+[AutoInject()]
+public interface ICreateQuestionRequestValidator : IValidator<CreateQuestionRequest>
+{
+  
```
Empty line

Levliv replied:
done

VLiamin commented on src/SurveyService.Data/GroupRepository.cs:
```diff
@@ -0,0 +1,29 @@
+
+    _provider.Groups.Add(dbGroup);
+    await _provider.SaveAsync();
+    return dbGroup.Id;
```
May be add emty line, like in others methods?

Levliv replied:
Done

Levliv committed:
Fixed style mistakes

maybee16 commented on src/SurveyService.Business/Commands/Group/Interfaces/ICreateGroupCommand.cs:
```diff
@@ -0,0 +1,11 @@
+
+public interface ICreateGroupCommand
+{
+  Task<OperationResultResponse<Guid>> ExecuteAsync(CreateGroupRequest request);
```
<Guid?>

Levliv replied:
Done

maybee16 commented on src/SurveyService.Business/Commands/Group/CreateGroupCommand.cs:
```diff
@@ -0,0 +1,59 @@
+    _validator = validator;
+  }
+
+  public async Task<OperationResultResponse<Guid>> ExecuteAsync(CreateGroupRequest request)
```
<Guid?>

Levliv replied:
Done

maybee16 commented on src/SurveyService.Business/Commands/Group/CreateGroupCommand.cs:
```diff
@@ -0,0 +1,59 @@
+        new List<string> { "Error occured while creating group." });
+    }
+    
+    return new OperationResultResponse<Guid>(body: (Guid)createGroupGuid);
```
Remove explicit type conversion (Guid)

Levliv replied:
Done

maybee16 commented on src/SurveyService.Mappers/Db/DbGroupMapper.cs:
```diff
@@ -0,0 +1,43 @@
+      ? null
+      : new DbGroup 
+      {
+        Id = groupId,
```
`Id = Guid.NewGuid();`

maybee16 replied:
Remove line 26

Levliv replied:
That won't work, as we have to pass the Id to mapper in this very statement

maybee16 commented on src/SurveyService.Mappers/Db/DbGroupMapper.cs:
```diff
@@ -0,0 +1,43 @@
+        CreatedBy = _httpContextAccessor.HttpContext.GetUserId(),
+        CreatedAtUtc = DateTime.Now,
+        ModifiedBy = null, 
+        ModifiedAtUtc = null,
```
May be remove 38-39 lines?

Levliv replied:
Done

Levliv committed:
Amended DbGroupMapper.cs useless = deleted

Levliv committed:
Return Group Guid as null if error occured.

Levliv committed:
Changed dateTime format to DateTime.UtcNow

MariyaSin commented on src/SurveyService.Models.Dto/Requests/Group/CreateGroupRequest.cs:
```diff
@@ -0,0 +1,13 @@
+{
+  public string Subject { get; set; }
+  public string Description { get; set; }
+  public bool IsActive { get; set; }
```
you can not create not active group, it doesn't make sense here, куьщму IsActive from create request

Levliv replied:
Done

MariyaSin commented on src/SurveyService.Models.Dto/Requests/Question/CreateQuestionRequest.cs:
```diff
@@ -0,0 +1,19 @@
+  public bool IsPrivate { get; set; }
+  public bool HasMultipleChoice { get; set; }
+  public bool HasCustomOptions { get; set; }
+  public bool IsActive { get; set; }
```
same as group

MariyaSin commented on src/SurveyService.Models.Dto/Requests/Group/CreateGroupRequest.cs:
```diff
@@ -0,0 +1,13 @@
+
+public record CreateGroupRequest
+{
+  public string Subject { get; set; }
```
add requred attribute 
this attribute guarantees that this property will not contain null

Levliv replied:
Done

MariyaSin commented on src/SurveyService.Models.Dto/Requests/Group/CreateGroupRequest.cs:
```diff
@@ -0,0 +1,13 @@
+  public string Description { get; set; }
+  public bool IsActive { get; set; }
+  
+  public ICollection<CreateQuestionRequest> Questions { get; set; }
```
add requred attribute
this will protect from the exception

Levliv replied:
Done

MariyaSin commented on src/SurveyService.Models.Dto/Requests/Question/CreateQuestionRequest.cs:
```diff
@@ -0,0 +1,19 @@
+  public bool IsPrivate { get; set; }
+  public bool HasMultipleChoice { get; set; }
+  public bool HasCustomOptions { get; set; }
+  public bool IsActive { get; set; }
```
same as group

Levliv replied:
I'd better just remove CreateQuetionRequest, this model someone else's responsibility.

MariyaSin commented on src/SurveyService.Models.Dto/Requests/Question/CreateQuestionRequest.cs:
```diff
@@ -0,0 +1,19 @@
+
+public record CreateQuestionRequest
+{
+  public string Content { get; set; }
```
requred attribute

Levliv replied:
I'd better just remove CreateQuetionRequest, this model someone else's responsibility.

MariyaSin commented on src/SurveyService.Validation/Group/CreateGroupRequestValidator.cs:
```diff
@@ -0,0 +1,31 @@
+      .MaximumLength(150)
+      .WithMessage("Subject should not exceed maximum length of 150 symbols.");
+
+    RuleFor(group => group.Description)
```
exception if description is null

Levliv replied:
For groups we allow empty description.
 Is it what you were asking about?

maybee16 replied:
`When (group => !string.IsNullOrEmpty(group.Description), () =>
{ 
RuleFor( ...)
}`

Levliv replied:
Done

MariyaSin commented on src/SurveyService.Validation/Group/Interfaces/ICreateGroupRequestValidator.cs:
```diff
@@ -0,0 +1,11 @@
+[AutoInject()]
+public interface ICreateGroupRequestValidator : IValidator<CreateGroupRequest>
+{
+  
```
remove empty line

Levliv replied:
Done

Levliv committed:
Fixed bugs in CreateGroupRequest.cs, fixed CreateQuestionRequest.cs, Added validation attributes for CreateGroupRequest.cs - cleaned up CreateGroupRequestValidator.cs

Levliv committed:
cleaned up CreateGroupRequestValidator.cs

Levliv committed:
Added HttpStatusCode to CreateGroupCommand.cs

Levliv committed:
Fixed CreateGroupRequestValidator.cs

maybee16 commented on src/SurveyService.Validation/Group/CreateGroupRequestValidator.cs:
```diff
@@ -14,10 +14,13 @@ public class CreateGroupRequestValidator : AbstractValidator<CreateGroupRequest>
-      .MaximumLength(500)
-      .WithMessage("Description should not exceed maximum length of 500 symbols.");
-
+    When(group => string.IsNullOrEmpty(group.Description), () =>
```
You forgot "!" before string :)

Levliv replied:
Yeah, straight to the point, thx.
Done

Levliv committed:
Fixed ! in CreateGroupRequestValidator.cs

maybee16 commented on src/SurveyService.Business/Commands/Group/CreateGroupCommand.cs:
```diff
@@ -0,0 +1,65 @@
+    {
+      return _responseCreator.CreateFailureResponse<Guid?>(
+        HttpStatusCode.BadRequest,
+        new List<string> { "Error occured while creating group." });
```
This error message is superfluous, it's created inside ResponseCreator

Levliv replied:
Done

Vasis3038 commented on src/SurveyService.Models.Dto/Requests/Group/CreateGroupRequest.cs:
```diff
@@ -0,0 +1,15 @@
+  public string Subject { get; set; }
+  public string Description { get; set; }
+
+  [Required(ErrorMessage = "At least 1 question is required.")]
```
Are you sure that an empty list won't pass validation?

Levliv replied:
Done

AlfaRedDog commented on src/SurveyService.Mappers/Db/Interfaces/IDbQuestionMapper.cs:
```diff
@@ -0,0 +1,10 @@
+
+public interface IDbQuestionMapper
+{
+  DbQuestion Map(CreateQuestionRequest dbQuestionRequest, Guid? groupId);
```
groupId is not a required parameter, it is included in CreateQuestionRequest

Levliv replied:
@vorobevamv 
I thought that CreateQuestionRequest doesn't include GroupId because, if we create just one question without a group we can't pass GroupId, so we set it to null in Mapper (we just can't pass GroupId). If we create questions while creating group we can't know GroupId as it doesn't exist yet.
Did I manage to clarify the situation?

vorobevamv replied:
Lev is right. GroupId in CreateQuestionRequest can be null, if it is a single question

Levliv committed:
Added empty list check in CreateGroupRequestValidator.cs,
refactor CreateGroupCommand.cs, Fixed Required attribute in CreateGroupRequest.cs

vorobevamv commented on src/SurveyService.Models.Dto/Requests/Group/CreateGroupRequest.cs:
```diff
@@ -0,0 +1,15 @@
+  [Required]
+  public string Subject { get; set; }
+  public string Description { get; set; }
+
```
remove empty line

Levliv replied:
Done

vorobevamv commented on src/SurveyService.Validation/Question/Interfaces/ICreateQuestionRequestValidator.cs:
```diff
@@ -0,0 +1,10 @@
+
+namespace LT.DigitalOffice.SurveyService.Validation.Question.Interfaces;
+
+[AutoInject()]
```
Don't need parentheses after attribute

Levliv replied:
Done

vorobevamv commented on src/SurveyService.Validation/Group/Interfaces/ICreateGroupRequestValidator.cs:
```diff
@@ -0,0 +1,10 @@
+
+namespace LT.DigitalOffice.SurveyService.Validation.Group.Interfaces;
+
+[AutoInject()]
```
remove parentheses

Levliv replied:
Done

vorobevamv commented on src/SurveyService.Validation/Group/CreateGroupRequestValidator.cs:
```diff
@@ -0,0 +1,32 @@
+    
+    RuleFor(group => group.Questions)
+      .Cascade(CascadeMode.Stop)
+      .Must(question => question.Any())
```
it is list, not single question => 'questions'

Levliv replied:
Done

vorobevamv commented on src/SurveyService.Validation/Group/CreateGroupRequestValidator.cs:
```diff
@@ -0,0 +1,32 @@
+      .Must(question => question.Any())
+      .WithMessage("At least 1 question in group is required.")
+      .ForEach(question => question.SetValidator(createQuestionRequestValidator))
+      .WithMessage("While Validating question error occured.");
```
don't need this line: errors from createQuestionRequestValidator would be added to error list

Levliv replied:
Done

vorobevamv commented on src/SurveyService.Validation/Group/CreateGroupRequestValidator.cs:
```diff
@@ -0,0 +1,32 @@
+      .Cascade(CascadeMode.Stop)
+      .Must(question => question.Any())
+      .WithMessage("At least 1 question in group is required.")
+      .ForEach(question => question.SetValidator(createQuestionRequestValidator))
```
please, move '.SetValidator(createQuestionRequestValidator))' to next line

Levliv replied:
Done

vorobevamv commented on src/SurveyService.Business/Commands/Group/Interfaces/ICreateGroupCommand.cs:
```diff
@@ -0,0 +1,11 @@
+
+namespace LT.DigitalOffice.SurveyService.Business.Commands.Group.Interfaces;
+
+public interface ICreateGroupCommand
```
you forgot something here... or in startup

Levliv replied:
Done

vorobevamv commented on src/SurveyService.Business/Commands/Group/CreateGroupCommand.cs:
```diff
@@ -0,0 +1,63 @@
+public class CreateGroupCommand : ICreateGroupCommand
+{
+  private readonly IGroupRepository _groupRepository;
+  private readonly IDbGroupMapper _dbGroupMapper;
```
it wouldn't work too

Levliv replied:
Done

vorobevamv commented on src/SurveyService.Business/Commands/Group/CreateGroupCommand.cs:
```diff
@@ -0,0 +1,63 @@
+
+public class CreateGroupCommand : ICreateGroupCommand
+{
+  private readonly IGroupRepository _groupRepository;
```
it wouldn't work; you need to add something (but not here)

Levliv replied:
Done

AlfaRedDog replied:
I ran your code, it still doesn't work. I have the same problem, [AutoInject] doesn't work and I don't know why

Levliv replied:
Have you implemented all the Interfaces (classes and methods) that are used?

AlfaRedDog replied:
Yes, all interfaces have an implementation. I took them from your pr's

AlfaRedDog replied:
You have everything working, just not enough implementations for the interfaces

vorobevamv commented on src/SurveyService.Models.Dto/Requests/Group/CreateGroupRequest.cs:
```diff
@@ -0,0 +1,15 @@
+  public string Description { get; set; }
+
+  [Required]
+  public ICollection<CreateQuestionRequest> Questions { get; set; }
```
Why not "List<CreateQuestionRequest>"?

Levliv replied:
[https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/guidelines-for-collections](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/guidelines-for-collections)

Levliv replied:
Maybe in this case it just doesn't make any difference?

vorobevamv replied:
maybe..
List is just more familiar)))

Levliv replied:
Ok, done

vorobevamv commented on src/SurveyService.Business/Commands/Group/CreateGroupCommand.cs:
```diff
@@ -0,0 +1,63 @@
+      );
+    }
+
+    OperationResultResponse<Guid?> response = new (){ Body = await _groupRepository.CreateAsync(_dbGroupMapper.Map(request)) };
```
OperationResultResponse<Guid?> response = new(body: await _groupRepository.CreateAsync(_dbGroupMapper.Map(request)));

(in one line)

Levliv replied:
Done

vorobevamv commented on src/SurveyService.Mappers/Db/DbGroupMapper.cs:
```diff
@@ -0,0 +1,41 @@
+public class DbGroupMapper : IDbGroupMapper
+{
+  private readonly IHttpContextAccessor _httpContextAccessor;
+  private readonly IDbQuestionMapper _questionMapper;
```
and this wouldn't work too

Levliv replied:
Done

vorobevamv commented on src/SurveyService.Mappers/Db/Interfaces/IDbQuestionMapper.cs:
```diff
@@ -0,0 +1,10 @@
+
+public interface IDbQuestionMapper
+{
+  DbQuestion Map(CreateQuestionRequest dbQuestionRequest, Guid? groupId);
```
CreateQuestionRequest createQuestionRequest

Levliv replied:
Done

vorobevamv commented:
Before making PR you should run your code to be sure it works. But code, which uses non-existing classes and methods, wouldn't work.
I'm talking about CreateQuestionRequest, CreateQuestionRequestValidator, DbQuestionMapper (may be I miss something).
I think, you have two ways: write them by yourself or wait, when they would be merged

Levliv committed:
Fixed mistakes in CreateGroupCommand.cs, CreateGroupRequest.cs, CreateGroupRequestValidator.cs, ICreateGroupCommand.cs, ICreateGroupRequestValidator.cs, ICreateQuestionRequestValidator.cs.
Added context accessor.

Levliv committed:
Refactor CreateGroupCommand.cs

Levliv committed:
Fixes typo in IDbQuestionMapper.cs

Levliv committed:
Fixed DI in IDbGroupMapper.cs, IDbQuestionMapper.cs, IGroupRepository.cs

Levliv commented:
> Before making PR you should run your code to be sure it works. But code, which uses non-existing classes and methods, wouldn't work. I'm talking about CreateQuestionRequest, CreateQuestionRequestValidator, DbQuestionMapper (may be I miss something). I think, you have two ways: write them by yourself or wait, when they would be merged

IMHO, it's better to wait for person, who's writing Question section to craete PR

Vasis3038 commented on src/SurveyService/Startup.cs:
```diff
@@ -85,6 +85,8 @@ public void ConfigureServices(IServiceCollection services)
       services.Configure<BaseServiceInfoConfig>(Configuration.GetSection(BaseServiceInfoConfig.SectionName));
       services.Configure<BaseRabbitMqConfig>(Configuration.GetSection(BaseRabbitMqConfig.SectionName));
 
+      services.AddHttpContextAccessor();
```
You already have it in line 71

Levliv replied:
Yeah, done

maybee16 committed:
Feature/create endpoint 'create' in option controller (#7)

* Created endpoint "Create" in OptionController

* Fixed CreateOptionCommand, OptionRepository, DbOptionMapper, CreateOptionRequestValidator

* Changed validator

* Created QuestionRepository and method GetAsync

* Created DoesExistAsync in QuestionRepository, fixed validator

* Fixed validator

* Fixed validator

* Deleted fluentvalidation

Levliv committed:
Fixed HttpAccessor

Levliv committed:
Fixed Startup.cs,
style changes with types in CreateGroupRequest.cs

AlfaRedDog committed:
Feature/create endpoint  create  in question controller (#9)

* add Create Question Endpoint

* delete InjectType

* Delete InjectType from ICreateQuestionCommand

* small fixes at QuestionRepository and CQRValidator

* fix Startup and Survey service references

* Fix Validator, renaming and refactoring

* Add single and group requests, and small fixes

* fix models CreateGroupQuestionRequest

* fix DbSingleQuestionMapper

* resolve QuestionRepository conflict

* fix DbSingleQuestionMapper, change CQRV

* Renaming and remove [Required]

* Add CreateQuestionOption methods, change DbGroup and migration, fix validator

* remove GetQuestionPropertiesFilter, small refactoring

* change question validator, remove group repo add getProperties method

* rework question create validator

* small refactoring

Levliv committed:
+ DbGroupQuestionMapper.cs, amended CreateGroupRequest.cs (added Deadline, added HasRealTimeResult), to fix: CreateGroupRequestValidator.cs, added newline CreateOptionRequestValidator.cs, fixed naming in CreateQuestionCommand.cs, - CreateQuestionRequest.cs, added newline and fixed naming CreateSingleQuestionRequestValidator.cs, fixed naming DbGroupMapper.cs, + DbGroupQuestionMapper.cs

Levliv committed:
Fixed style DbGroupQuestionMapper.cs

Levliv committed:
Fixed typing CreateGroupQuestionRequest.cs,
added CreateGroupRequestValidator.cs,
amended CreateGroupRequestValidator.cs (check that group contains at least 1 question, check deadline for group, added validation procedure for each question)
style DbGroupQuestionMapper.cs,
fixed typing DbSingleQuestionMapper.cs,

Levliv committed:
Style changes

Levliv committed:
Fixed

Levliv committed:
Merge branch 'develop' into feature/CreateGroup_Endpoint

Vasis3038 commented on src/SurveyService.Business/Commands/Question/CreateQuestionCommand.cs:
```diff
@@ -19,14 +19,13 @@ public class CreateQuestionCommand : ICreateQuestionCommand
-  private readonly ICreateQuestionRequestValidator _validator;
-  private readonly IResponseCreator _responseCreator;
+  private readonly ICreateSingleQuestionRequestValidator _validator;
 
```
you must have accidentally removed _responseCreator

Levliv replied:
Fixed

AlfaRedDog commented on src/SurveyService.Business/Commands/Group/CreateGroupCommand.cs:
```diff
@@ -0,0 +1,62 @@
+      );
+    }
+
+    OperationResultResponse<Guid?> response = new (body: await _groupRepository.CreateAsync(_dbGroupMapper.Map(request)));
```
"body:..." should be on the next line

Levliv replied:
Fixed

AlfaRedDog commented on src/SurveyService.Mappers/Db/DbGroupQuestionMapper.cs:
```diff
@@ -0,0 +1,38 @@
+using LT.DigitalOffice.SurveyService.Models.Dto.Requests.Question;
+using Microsoft.AspNetCore.Mvc;
+using System;
+namespace LT.DigitalOffice.SurveyService.Mappers.Db;
```
add empty line before line 6

Levliv replied:
Fixed

AlfaRedDog reviewed (commented):
also you should delete ICreateQuestionRequestValidator and CreateQuestionRequestValidator

Levliv committed:
Fixed NRE DbSingleQuestionMapper.cs, see https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/member-access-operators#null-conditional-operators--and-

Levliv committed:
Merge remote-tracking branch 'origin/feature/CreateGroup_Endpoint' into feature/CreateGroup_Endpoint

Levliv committed:
Fixed style CreateGroupQuestionRequestValidator.cs

Levliv committed:
Fixed CreateQuestionCommand.cs

Levliv committed:
Fixed CreateGroupCommand.cs, fixed DbGroupQuestionMapper.cs

Levliv committed:
Fixed minor style changes CreateOptionRequestValidator.cs, DbGroupQuestionMapper.cs, deleted CreateQuestionRequestValidator.cs, deleted ICreateQuestionRequestValidator.cs

vorobevamv commented on src/SurveyService.Models.Dto/Requests/Group/CreateGroupRequest.cs:
```diff
@@ -0,0 +1,17 @@
+  public string Subject { get; set; }
+  public string Description { get; set; }
+  public DateTime? Deadline { get; set; }
+  public bool HasRealTimeResults { get; set; }
```
HasRealTimeResult

Levliv replied:
Fixed

vorobevamv commented on src/SurveyService.Mappers/Db/DbGroupMapper.cs:
```diff
@@ -0,0 +1,41 @@
+        IsActive = true,
+        CreatedBy = _httpContextAccessor.HttpContext.GetUserId(),
+        CreatedAtUtc = DateTime.UtcNow,
+        Questions = request.Questions.Select(question => _groupQuestionMapper.Map(question, groupId, request.Deadline, request.HasRealTimeResults)).ToList()
```
It's very difficult schema with IDbGroupQuestionMapper and IDbQuestionOptionMapper; also they are used only here.
I think, better is to remove IDbGroupQuestionMapper, DbGroupQuestionMapper, IDbQuestionOptionMapper, DbQuestionOptionMapper and write all mapping here or overload Map in DbOptionMapper and DbQuestionMapper

Levliv replied:
BTW, IDbQuestionOptionMapper is used in many other cases. The idea was to make clear how we map questions, options and groups.

vorobevamv replied:
Look, there is method Map in DbOptionMapper, that maps and returns DbOption;
DbQuestionOptionMapper does the same; don't need DbQuestionOptionMapper - overload method Map in DbOptionMapper;
with question the same

Levliv replied:
Got it, done

vorobevamv commented on src/SurveyService.Validation/Group/CreateGroupRequestValidator.cs:
```diff
@@ -0,0 +1,40 @@
+public class CreateGroupRequestValidator : AbstractValidator<CreateGroupRequest>, ICreateGroupRequestValidator
+{
+  public CreateGroupRequestValidator(
+    [FromServices] ICreateGroupQuestionRequestValidator createGroupQuestionRequestValidator)
```
don't need "[FromServices]" here; it's not wrong, but unnecessary

Levliv replied:
Fixed

vorobevamv commented on src/SurveyService.Validation/Question/CreateGroupQuestionRequestValidator.cs:
```diff
@@ -0,0 +1,28 @@
+
+namespace LT.DigitalOffice.SurveyService.Validation.Question;
+
+public class CreateGroupQuestionRequestValidator : AbstractValidator<CreateGroupQuestionRequest>, ICreateGroupQuestionRequestValidator
```
you use this validator just once in CreateGroupRequestValidator; what's the point to separate it?

Levliv replied:
Just for readability

vorobevamv commented on src/SurveyService.Validation/Group/CreateGroupRequestValidator.cs:
```diff
@@ -0,0 +1,40 @@
+
+public class CreateGroupRequestValidator : AbstractValidator<CreateGroupRequest>, ICreateGroupRequestValidator
+{
+  public CreateGroupRequestValidator(
```
please, add check:
if Deadline is null, HasRealTimeResult is true

Levliv replied:
Shouldn't we add the same for CreateSingleQuestionRequestValidator?

vorobevamv replied:
should. i forgot((

Levliv replied:
Done

Levliv committed:
Fixed CreateGroupRequestValidator.cs, Fixed CreateGroupRequest.cs, Fixed DbGroupMapper.cs

Levliv committed:
Style CreateGroupRequestValidator.cs

Levliv committed:
Style

Levliv committed:
style

Levliv committed:
Fixed CreateGroupRequestValidator.cs, fixed CreateSingleQuestionRequestValidator.cs

Levliv committed:
Fixed CreateSingleQuestionRequestValidator.cs

Levliv committed:
Overloaded Map in DbOptionMapper.cs insead of using DbQuestionOptionMapper.cs

Levliv committed:
Map overloaded in DbQuestionMapper.cs, deleted GroupQuestionMapper
